www/js/lib/popovers.js

```diff
@@ -0,0 +1,455 @@
+            font-size: 20px;
+            font-family: sans-serif;
+        }
+        
+        #popcloseicon:hover { 
+            cursor: pointer;
+        }
+        
+        #popbreakouticon {
+            height: 18px;
+            margin-right: 18px;
+        }
+        
+        #popbreakouticon:hover {
+            cursor: pointer;
+        }
+        
+        /* Prevent native iOS popover on Safari if option is enabled */
+        body { 
+            -webkit-touch-callout: none !important;
+        }
+        `,
+    // The id of the style element for easy manipulation
+    'kiwixtooltipstylesheet');
+}
+
+/**
+ * Attaches a popover div for the given link to the given document's DOM
+ * @param {Event} ev The event which has fired this popover action
+ * @param {Element} link The link element that is being actioned
+ * @param {Object} appstate The globlal object defined in app.js that holds the current state of the app
+ * @param {Boolean} dark An optional value to switch colour theme to dark if true
+ * @param {ZIMArchive} archive The archive from which the popover information is extracted
+ * @returns {Promise<div>} A Promise for the attached popover div or undefined if the popover is not attached
+ */
+function populateKiwixPopoverDiv (ev, link, appstate, dark, archive) {
+    // Do not show popover if the user has initiated an article load (set in filterClickEvent)
+    if (link.articleisloading || link.popoverisloading) return Promise.resolve();
+    const linkHref = link.getAttribute('href');
+    // Do not show popover if there is no href or with certain landing pages
+    if (!linkHref || /^wikivoyage/i.test(archive.file.name) &&
+      (appstate.expectedArticleURLToBeDisplayed === archive.landingPageUrl ||
+      appstate.expectedArticleURLToBeDisplayed === 'A/Wikivoyage:Offline_reader_Expedition/Home_page')) {
+        return Promise.resolve();
+    }
+    link.popoverisloading = true;
+    // Do not display a popover if one is already showing for the current link
+    const kiwixPopover = ev.target.ownerDocument.querySelector('.kiwixtooltip');
+    if (kiwixPopover && kiwixPopover.dataset.href === linkHref) return Promise.resolve();
+    // console.debug('Attaching popover...');
+    const currentDocument = ev.target.ownerDocument;
+    const articleWindow = currentDocument.defaultView;
+    // Remove any existing popover(s) that the user may not have closed before creating a new one
+    removeKiwixPopoverDivs(currentDocument);
+    setTimeout(function () {
+        // Check if the user has moved away from the link or has clicked it, and abort display of popover if so
+        if (link.articleisloading || !link.matches(':hover') && !link.touched && currentDocument.activeElement !== link) {
+            // console.debug('Aborting popover display for ' + linkHref + ' because user has moved away from link or clicked it');
+            link.popoverisloading = false;
+            return Promise.resolve();
```

Comment:
I don't think this does what you think it does. You're return a promise from the anonymous function inside of the setTimeout?